src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,505 @@
+                StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
+                return ParseSequenceWorker();
+            }
+            finally
+            {
+                _recursionDepth--;
+            }
+        }
+
+        private ImmutableArray<JsonValueNode> ParseSequenceWorker()
+        {
+            using var _ = ArrayBuilder<JsonValueNode>.GetInstance(out var result);
+
+            while (ShouldConsumeSequenceElement())
+                result.Add(ParseValue());
+
+            return result.ToImmutable();
+        }
+
+        private bool ShouldConsumeSequenceElement()
+        {
+            if (_currentToken.Kind == JsonKind.EndOfFile)
+            {
+                return false;
+            }
+
+            if (_currentToken.Kind == JsonKind.CloseBraceToken)
+            {
+                return !_inObject;
+            }
+
+            if (_currentToken.Kind == JsonKind.CloseBracketToken)
+            {
+                return !_inArray;
+            }
+
+            if (_currentToken.Kind == JsonKind.CloseParenToken)
+            {
+                return !_inConstructor;
+            }
+
+            return true;
+        }
+
+        private JsonValueNode ParseValue()
+            => _currentToken.Kind switch
+            {
+                JsonKind.OpenBraceToken => ParseObject(),
+                JsonKind.OpenBracketToken => ParseArray(),
+                JsonKind.CommaToken => ParseCommaValue(),
+                _ => ParseLiteralOrPropertyOrConstructor(),
+            };
+
+        private static void SplitLiteral(JsonToken literalToken, out JsonToken minusToken, out JsonToken newLiteralToken)
+        {
+            minusToken = CreateToken(
+                JsonKind.MinusToken, literalToken.LeadingTrivia,
+                literalToken.VirtualChars.GetSubSequence(new TextSpan(0, 1)),
+                ImmutableArray<JsonTrivia>.Empty);
+            newLiteralToken = CreateToken(
+                literalToken.Kind,
+                ImmutableArray<JsonTrivia>.Empty,
+                literalToken.VirtualChars.GetSubSequence(TextSpan.FromBounds(1, literalToken.VirtualChars.Length)),
+                literalToken.TrailingTrivia,
+                literalToken.Diagnostics);
+        }
+
+        private JsonPropertyNode ParseProperty(JsonToken stringLiteralOrText)
+        {
+            Debug.Assert(_currentToken.Kind == JsonKind.ColonToken);
+            if (stringLiteralOrText.Kind != JsonKind.StringToken)
+            {
+                stringLiteralOrText = stringLiteralOrText.With(kind: JsonKind.TextToken);
+            }
+
+            var colonToken = ConsumeCurrentToken();
+            // Newtonsoft allows "{ a: , }" as a legal property.
+            if (_currentToken.Kind == JsonKind.CommaToken)
+            {
+                return new JsonPropertyNode(
+                    stringLiteralOrText, colonToken,
+                    new JsonCommaValueNode(CreateMissingToken(JsonKind.CommaToken)));
```

Comment:
Trying to reconcile the text with the resulting parse here -- so the property is valid, but we count the value as being a missing comma? How would something like `{ a: , , }` parse?

Reply:
That will be a property whose value is a comma, followed by a comma value in the property list itself.  will ensure we have test fort his.

Reply:
so this will be a property, followed by two comma values.  sigh :(

Reply:
test ensured.